examples/erc1155/lib.rs

```diff
@@ -0,0 +1,687 @@
+            );
+
+            if let Some(b) = self.balances.get_mut(&(from, token_id)) {
+                *b -= value
+            }
+
+            self.balances
+                .entry((to, token_id))
+                .and_modify(|b| *b += value)
+                .or_insert(value);
+
+            self.env().emit_event(TransferSingle {
+                operator: self.env().caller(),
+                from,
+                to,
+                token_id,
+                value,
+            });
+
+            // Quick Haxx, otherwise my tests just panic due to the use of eval_contract()
+            #[cfg(not(test))]
+            {
+                // If our recipient is a smart contract we need to see if they accept or
+                // reject this transfer. If they reject it we need to revert the call.
+                let params = build_call::<ink_env::DefaultEnvironment>()
+                    .callee(to)
+                    .gas_limit(5000) // Q: What's the correct amount to use here?
+                    .exec_input(
+                        ExecutionInput::new(Selector::new(MAGIC_VALUE))
+                        .push_arg(self.env().caller())
+                        .push_arg(from)
+                        .push_arg(token_id)
+                        .push_arg(value)
+                        .push_arg(data)
+                    )
+                    .returns::<ReturnType<Vec<u8>>>()
+                    .params();
+
+                match ink_env::eval_contract(&params) {
+                    Ok(v) => {
+                        assert_eq!(
+                            v,
+                            &MAGIC_VALUE[..],
+                            "Recipient contract does not accept token transfers."
+                        )
+                    }
+                    Err(e) => {
+                        match e {
+                            ink_env::Error::CodeNotFound => {
+                                // Our recipient wasn't a smart contract, so there's nothing more for
+                                // us to do
+                            }
+                            _ => {
+                                // I have no insight to what's happening here, I can't deploy
+                                // the following:
+                                //
+                                // ink_env::debug_println(&ink_prelude::format!("{:?}", e));
+                                // panic!("{:?}", e)
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    impl super::Erc1155 for Contract {
+        #[ink(message)]
+        fn safe_transfer_from(
+            &mut self,
+            from: AccountId,
+            to: AccountId,
+            token_id: TokenId,
+            value: Balance,
+            data: Vec<u8>,
+        ) {
+            // Q: Does the caller change if the function is called from within this smart contract?
```

Comment:
A: No

:)

Reply:
Interesting question though ... never thought about this potential confusion before.